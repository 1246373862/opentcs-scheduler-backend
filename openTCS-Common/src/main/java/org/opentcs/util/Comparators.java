/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.util;

import java.util.Comparator;
import org.opentcs.data.TCSObject;
import org.opentcs.data.TCSObjectReference;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.util.annotations.ScheduledApiChange;

/**
 * Some commonly used comparator implementations.
 */
public final class Comparators {

  /**
   * Prevents undesired instantiation.
   */
  private Comparators() {
  }

  /**
   * Returns a comparator for sorting <code>TCSObject</code>s lexicographically by their names.
   *
   * @return A comparator for sorting <code>TCSObject</code>s lexicographically by their names.
   */
  public static Comparator<TCSObject<?>> objectsByName() {
    return Comparator.comparing(TCSObject::getName);
  }

  /**
   * Returns a comparator for sorting <code>TCSObjectReference</code>s lexicographically by their
   * names.
   *
   * @return A comparator for sorting <code>TCSObjectReference</code>s lexicographically by their
   * names.
   */
  public static Comparator<TCSObjectReference<?>> referencesByName() {
    return Comparator.comparing(TCSObjectReference::getName);
  }

  /**
   * A comparator for sorting transport orders by their deadlines, with the most urgent ones coming
   * first.
   * Transport orders are sorted by their deadlines first; if two orders have exactly the same
   * deadline, they are sorted by their (unique) creation times, with the older one coming first.
   *
   * @return A comparator for sorting transport orders by their deadlines.
   */
  public static Comparator<TransportOrder> ordersByDeadline() {
    return Comparator.comparing(TransportOrder::getDeadline).thenComparing(ordersByAge());
  }

  /**
   * A comparator for sorting transport orders by their age, with the oldest ones coming first.
   *
   * @return A comparator for sorting transport orders by their age.
   */
  public static Comparator<TransportOrder> ordersByAge() {
    return Comparator.comparing(TransportOrder::getCreationTime).thenComparing(objectsByName());
  }

  /**
   * A comparator for sorting peripheral jobs by their age, with the oldest ones coming first.
   *
   * @return A comparator for sorting peripheral jobs by their age.
   */
  public static Comparator<PeripheralJob> jobsByAge() {
    return Comparator.comparing(PeripheralJob::getCreationTime).thenComparing(objectsByName());
  }

  /**
   * A comparator for ordering ModelLayoutElements by the names of their
   * visualized objects.
   *
   * @return A comparator for ordering ModelLayoutElements by the names of their
   * visualized objects.
   */
  @Deprecated
  @ScheduledApiChange(details = "Will be removed.", when = "6.0")
  public static Comparator<org.opentcs.data.model.visualization.ModelLayoutElement>
      modelLayoutElementsByName() {
    return (org.opentcs.data.model.visualization.ModelLayoutElement o1,
        org.opentcs.data.model.visualization.ModelLayoutElement o2)
        -> o1.getVisualizedObject().getName().compareTo(
            o2.getVisualizedObject().getName());
  }
}
