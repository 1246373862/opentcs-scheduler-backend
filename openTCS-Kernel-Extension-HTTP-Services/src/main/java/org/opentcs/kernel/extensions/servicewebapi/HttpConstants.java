/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
