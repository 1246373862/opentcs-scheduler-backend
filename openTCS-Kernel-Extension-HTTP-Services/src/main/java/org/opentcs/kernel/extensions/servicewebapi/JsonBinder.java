/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;

/**
 * Binds JSON strings to objects and vice versa.
 */
public class JsonBinder {

  /**
   * Maps between objects and their JSON representations.
   */
  private final ObjectMapper objectMapper
      = new ObjectMapper()
          .registerModule(new JavaTimeModule())
          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

  /**
   * Creates a new instance.
   */
  public JsonBinder() {
  }

  /**
   * Maps the given JSON string to an object.
   *
   * @param <T> The type of object to map to.
   * @param jsonString The JSON string.
   * @param clazz The type of object to map to.
   * @return The object created from the JSON string.
   * @throws IllegalArgumentException In case there was a problem mapping the given object from
   * JSON.
   * (An IllegalArgumentException is mapped to HTTP status code 400, indicating a client error.)
   */
  public <T> T fromJson(String jsonString, Class<T> clazz)
      throws IllegalArgumentException {
    try {
      return objectMapper.readValue(jsonString, clazz);
    }
    catch (IOException exc) {
      throw new IllegalArgumentException("Could not parse JSON input", exc);
    }
  }

  /**
   * Maps the given object to a JSON string.
   *
   * @param object The object to be mapped.
   * @return The JSON string representation of the object.
   * @throws IllegalStateException In case there was a problem mapping the given object to JSON.
   * (An IllegalStateException is mapped to HTTP status code 500, indicating an internal error.)
   */
  public String toJson(Object object)
      throws IllegalStateException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(object);
    }
    catch (JsonProcessingException exc) {
      throw new IllegalStateException("Could not produce JSON output", exc);
    }
  }

  /**
   * Maps the given throwable to a JSON string.
   *
   * @param t The throwable to be mapped.
   * @return A JSON string for the given throwable, consisting of a single-element array containing
   * the throwable's message.
   * @throws IllegalStateException In case there was a problem mapping the given object to JSON.
   * (An IllegalStateException is mapped to HTTP status code 500, indicating an internal error.)
   */
  public String toJson(Throwable t)
      throws IllegalStateException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(objectMapper.createArrayNode().add(t.getMessage()));
    }
    catch (JsonProcessingException exc) {
      throw new IllegalStateException("Could not produce JSON output", exc);
    }
  }

}
