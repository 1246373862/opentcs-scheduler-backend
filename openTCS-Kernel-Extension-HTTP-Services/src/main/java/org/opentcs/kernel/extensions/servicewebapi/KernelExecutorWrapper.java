/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import javax.inject.Inject;
import org.opentcs.access.KernelRuntimeException;
import org.opentcs.customizations.kernel.KernelExecutor;

/**
 * Calls callables/runnables via the kernel executor and waits for the outcome.
 */
public class KernelExecutorWrapper {

  private final ExecutorService kernelExecutor;

  /**
   * Creates a new instance.
   *
   * @param kernelExecutor The kernel executor.
   */
  @Inject
  public KernelExecutorWrapper(@KernelExecutor ExecutorService kernelExecutor) {
    this.kernelExecutor = requireNonNull(kernelExecutor, "kernelExecutor");
  }

  /**
   * Calls the given callable via the kernel executor and waits for the outcome.
   *
   * @param <T> The callable's return type.
   * @param callable The callable.
   * @return The result of the call.
   * @throws IllegalStateException In case the call via the kernel executor was unexpectedly
   * interrupted.
   * @throws RuntimeException In case an exception was thrown from the callable. If the exception
   * thrown is a {@code RuntimeException}, it is forwarded directly; if it is not a
   * {@code RuntimeException}, it is wrapped in a {@link KernelRuntimeException}.
   */
  public <T> T callAndWait(Callable<T> callable)
      throws IllegalStateException, RuntimeException {
    requireNonNull(callable, "callable");

    try {
      return kernelExecutor.submit(callable).get();
    }
    catch (InterruptedException exc) {
      throw new IllegalStateException("Unexpectedly interrupted");
    }
    catch (ExecutionException exc) {
      if (exc.getCause() instanceof RuntimeException) {
        throw (RuntimeException) exc.getCause();
      }
      throw new KernelRuntimeException(exc.getCause());
    }
  }

  /**
   * Calls the given runnable via the kernel executor and waits for the outcome.
   *
   * @param runnable The runnable.
   * @throws IllegalStateException In case the call via the kernel executor was unexpectedly
   * interrupted.
   * @throws RuntimeException In case an exception was thrown from the runnable. If the exception
   * thrown is a {@code RuntimeException}, it is forwarded directly; if it is not a
   * {@code RuntimeException}, it is wrapped in a {@link KernelRuntimeException}.
   */
  public void callAndWait(Runnable runnable)
      throws IllegalStateException, RuntimeException {
    requireNonNull(runnable, "runnable");

    callAndWait(Executors.callable(runnable));
  }
}
