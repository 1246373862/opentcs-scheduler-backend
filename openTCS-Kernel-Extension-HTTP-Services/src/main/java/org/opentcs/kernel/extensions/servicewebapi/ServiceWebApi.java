/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import com.google.common.util.concurrent.Uninterruptibles;
import static java.util.Objects.requireNonNull;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import org.opentcs.access.KernelRuntimeException;
import org.opentcs.access.SslParameterSet;
import org.opentcs.components.kernel.KernelExtension;
import org.opentcs.data.ObjectExistsException;
import org.opentcs.data.ObjectUnknownException;
import org.opentcs.kernel.extensions.servicewebapi.v1.V1RequestHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import spark.Service;

/**
 * Provides an HTTP interface for basic administration needs.
 */
public class ServiceWebApi
    implements KernelExtension {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(ServiceWebApi.class);
  /**
   * The interface configuration.
   */
  private final ServiceWebApiConfiguration configuration;
  /**
   * Authenticates incoming requests.
   */
  private final Authenticator authenticator;
  /**
   * Handles requests for API version 1.
   */
  private final V1RequestHandler v1RequestHandler;
  /**
   * Binds JSON data to objects and vice versa.
   */
  private final JsonBinder jsonBinder;
  /**
   * The connection encryption configuration.
   */
  private final SslParameterSet sslParamSet;
  /**
   * The actual HTTP service.
   */
  private Service service;
  /**
   * Whether this kernel extension is initialized.
   */
  private boolean initialized;

  /**
   * Creates a new instance.
   *
   * @param configuration The interface configuration.
   * @param sslParamSet The SSL parameter set.
   * @param authenticator Authenticates incoming requests.
   * @param jsonBinder Binds JSON data to objects and vice versa.
   * @param v1RequestHandler Handles requests for API version 1.
   */
  @Inject
  public ServiceWebApi(ServiceWebApiConfiguration configuration,
                       SslParameterSet sslParamSet,
                       Authenticator authenticator,
                       JsonBinder jsonBinder,
                       V1RequestHandler v1RequestHandler) {
    this.configuration = requireNonNull(configuration, "configuration");
    this.sslParamSet = requireNonNull(sslParamSet, "sslParamSet");
    this.authenticator = requireNonNull(authenticator, "authenticator");
    this.jsonBinder = requireNonNull(jsonBinder, "jsonBinder");
    this.v1RequestHandler = requireNonNull(v1RequestHandler, "v1RequestHandler");
  }

  @Override
  public void initialize() {
    if (isInitialized()) {
      return;
    }

    v1RequestHandler.initialize();

    service = Service.ignite()
        .ipAddress(configuration.bindAddress())
        .port(configuration.bindPort());

    if (configuration.useSsl()) {
      service.secure(sslParamSet.getKeystoreFile().getAbsolutePath(),
                     sslParamSet.getKeystorePassword(),
                     null,
                     null);
    }
    else {
      LOG.warn("Encryption disabled, connections will not be secured!");
    }

    service.before((request, response) -> {
      if (!authenticator.isAuthenticated(request)) {
        // Delay the response a bit to slow down brute force attacks.
        Uninterruptibles.sleepUninterruptibly(2, TimeUnit.SECONDS);
        service.halt(403, "Not authenticated.");
      }

      // Add a CORS header to allow cross-origin requests from all hosts.
      // This also makes using the "try it out" buttons in the Swagger UI documentation possible.
      response.header("Access-Control-Allow-Origin", "*");
    });

    // Reflect that we allow cross-origin requests for any headers and methods.
    service.options(
        "/*",
        (request, response) -> {
          String accessControlRequestHeaders = request.headers("Access-Control-Request-Headers");
          if (accessControlRequestHeaders != null) {
            response.header("Access-Control-Allow-Headers", accessControlRequestHeaders);
          }

          String accessControlRequestMethod = request.headers("Access-Control-Request-Method");
          if (accessControlRequestMethod != null) {
            response.header("Access-Control-Allow-Methods", accessControlRequestMethod);
          }

          return "OK";
        });

    // Register routes for API versions here.
    service.path("/v1", () -> v1RequestHandler.addRoutes(service));

    service.exception(IllegalArgumentException.class, (exception, request, response) -> {
                    response.status(400);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(ObjectUnknownException.class, (exception, request, response) -> {
                    response.status(404);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(ObjectExistsException.class, (exception, request, response) -> {
                    response.status(409);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(KernelRuntimeException.class, (exception, request, response) -> {
                    response.status(500);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(IllegalStateException.class, (exception, request, response) -> {
                    response.status(500);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });

    initialized = true;
  }

  @Override
  public void terminate() {
    if (!isInitialized()) {
      return;
    }

    v1RequestHandler.terminate();
    service.stop();

    initialized = false;
  }

  @Override
  public boolean isInitialized() {
    return initialized;
  }

}
