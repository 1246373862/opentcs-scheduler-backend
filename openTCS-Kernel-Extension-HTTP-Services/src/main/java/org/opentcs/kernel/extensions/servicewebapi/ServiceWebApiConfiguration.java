/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import org.opentcs.configuration.ConfigurationEntry;
import org.opentcs.configuration.ConfigurationPrefix;

/**
 * Configuration entries for the service web API.
 */
@ConfigurationPrefix(ServiceWebApiConfiguration.PREFIX)
public interface ServiceWebApiConfiguration {

  /**
   * The prefix for all configuration entries here.
   */
  String PREFIX = "servicewebapi";

  @ConfigurationEntry(
      type = "Boolean",
      description = "Whether to enable the interface.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "0"
  )
  boolean enable();

  @ConfigurationEntry(
      type = "IP address",
      description = "Address to which to bind the HTTP server, e.g. 0.0.0.0 or 127.0.0.1.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "1"
  )
  String bindAddress();

  @ConfigurationEntry(
      type = "Integer",
      description = "Port to which to bind the HTTP server.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "2"
  )
  int bindPort();

  @ConfigurationEntry(
      type = "String",
      description = "Key allowing access to the API.",
      changesApplied = ConfigurationEntry.ChangesApplied.INSTANTLY,
      orderKey = "3"
  )
  String accessKey();

  @ConfigurationEntry(
      type = "Integer",
      description = "Maximum number of status events to be kept.",
      changesApplied = ConfigurationEntry.ChangesApplied.INSTANTLY,
      orderKey = "4"
  )
  int statusEventsCapacity();

  @ConfigurationEntry(
      type = "Boolean",
      description = "Whether to use SSL to encrypt connections.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "5"
  )
  boolean useSsl();
}
