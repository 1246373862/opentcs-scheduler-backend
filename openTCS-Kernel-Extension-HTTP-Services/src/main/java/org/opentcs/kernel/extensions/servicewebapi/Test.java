package org.opentcs.kernel.extensions.servicewebapi;

import org.opentcs.components.kernel.KernelExtension;

public class Test implements KernelExtension {

  private volatile boolean enabled;

  @Override
  public void initialize() {
    if (enabled){
      return;
    }
    System.out.println("初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~");
  }

  @Override
  public boolean isInitialized() {
    return enabled;
  }

  @Override
  public void terminate() {
    if (!enabled) {
      return;
    }
    enabled = false;
    System.out.println("结束！！");
  }
}
