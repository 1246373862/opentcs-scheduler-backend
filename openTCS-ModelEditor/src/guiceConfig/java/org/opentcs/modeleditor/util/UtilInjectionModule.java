/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.modeleditor.util;

import com.google.inject.AbstractModule;
import javax.inject.Singleton;
import org.opentcs.guing.common.util.PanelRegistry;

/**
 * A default Guice module for this package.
 */
public class UtilInjectionModule
    extends AbstractModule {

  /**
   * Creates a new instance.
   */
  public UtilInjectionModule() {
  }

  @Override
  protected void configure() {
    bind(PanelRegistry.class).in(Singleton.class);
  }
}
