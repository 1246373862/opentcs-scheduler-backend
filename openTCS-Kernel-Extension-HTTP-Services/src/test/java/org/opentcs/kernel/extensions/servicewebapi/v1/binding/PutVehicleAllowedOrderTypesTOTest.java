/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import java.util.List;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;

/**
 */
public class PutVehicleAllowedOrderTypesTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  public void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  public void jsonSample() {
    PutVehicleAllowedOrderTypesTO to
        = new PutVehicleAllowedOrderTypesTO(List.of("some-orderType",
                                                   "another-orderType",
                                                   "orderType-3"));

    Approvals.verify(jsonBinder.toJson(to));
  }
}
