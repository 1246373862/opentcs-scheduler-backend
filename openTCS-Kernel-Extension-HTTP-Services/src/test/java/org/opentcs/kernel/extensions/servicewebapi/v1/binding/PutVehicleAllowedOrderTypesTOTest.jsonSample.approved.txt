{
  "orderTypes" : [ "some-orderType", "another-orderType", "orderType-3" ]
}